pymc3/distributions/mixture.py

```diff
@@ -612,3 +618,183 @@ def __init__(self, w, mu, sigma=None, tau=None, sd=None, comp_shape=(), *args, *
 
     def _distr_parameters_for_repr(self):
         return ["w", "mu", "sigma"]
+
+
+class MixtureSameFamily(Distribution):
+    R"""
+    Mixture log-likelihood
+
+    .. math::f(x \mid w, \theta) = \sum_{i = 1}^n w_i f_i(x \mid \theta_i)\textrm{ Along mixture\_axis}
+
+    ========  ============================================
+    Support   :math:`\textrm{support}(f)`
+    Mean      :math:`w\mu`
+    ========  ============================================
+
+    Parameters
+    ----------
+    w: array of floats
+        w >= 0 and w <= 1
+        the mixture weights
+    comp_dists: multidimensional PyMC3 distribution (e.g. `pm.Multinomial.dist(...)`)
+        with shape (batch_shape, mixture_axis, event_shape)
```

Comment:
Yes, pymc3 makes no distinction between `batch_shape` and `event_shape`. Moreover, pymc3 doesn't consistently reduce the `event_shape` dimensions in the `logp` calculations because it just completely reduces all axis in the model's `logp`. This prevents pymc3 from being able to vectorize its step functions across chains, as TFP does.
For someone that doesn't know TFP or didn't have anything to do with pymc4, the batch and event shape will be completely foreign concepts. I don't know if we should explain what they are, we should simply emphasize that the shape of the component distribution passed to the `MixtureSameFamily` has an axis that the user can identify as the `mixture_axis`. We could write it down like `(i_0, ..., i_n, mixture_axis, i_n+1, ..., i_N)`. And we should also say that the `mixture_axis` will be "consumed" by the mixture distribution, so the user should be aware that its mixture will end up with a shape like `(i_0, ..., i_n, i_n+1, ..., i_N)`.